sys/include/xtimer/implementation.h

```diff
@@ -111,22 +113,24 @@ void _xtimer_tsleep(uint32_t offset, uint32_t long_offset);
 
 static inline uint32_t _xtimer_now(void)
 {
-#if XTIMER_MASK
-    uint32_t latched_high_cnt, now;
-
-    /* _high_cnt can change at any time, so check the value before
-     * and after reading the low-level timer. If it hasn't changed,
-     * then it can be safely applied to the timer count. */
-
-    do {
-        latched_high_cnt = _xtimer_high_cnt;
-        now = _xtimer_lltimer_now();
-    } while (_xtimer_high_cnt != latched_high_cnt);
+    uint32_t now, elapsed;
 
-    return latched_high_cnt | now;
+    /* time sensitive since _long_cnt and _xtimer_high_cnt are updated here */
+    uint8_t state = irq_disable();
+    now = _xtimer_lltimer_now();
+#if XTIMER_MASK
+    elapsed = now - _xtimer_lltimer_mask((uint32_t)_xtimer_current_time);
+    _xtimer_current_time += (uint64_t)elapsed;
+    _xtimer_high_cnt = ((uint32_t)_xtimer_current_time) & XTIMER_MASK;
+    _long_cnt = (uint32_t)(_xtimer_current_time >> 32);
 #else
-    return _xtimer_lltimer_now();
+    elapsed = now - ((uint32_t)_xtimer_current_time & 0xFFFFFFFF);
+    _xtimer_current_time += (uint64_t)elapsed;
+    _long_cnt = (uint32_t)(_xtimer_current_time >> 32);
 #endif
+    irq_restore(state);
+
+    return (uint32_t)_xtimer_current_time;
```

Comment:
This function updates "64-bit absolute current time" (_xtimer_current_time). For safe update, this function calculates how much time elapsed from the last _xtimer_current_time update (i.e., offset). Other important parameters, _long_cnt and _xtimer_high_cnt, all come out of _xtimer_current_time.